### PR TITLE
Add interpreter for BatchNormInferenceOp

### LINKED ISSUE #963: Add interpreter for BatchNormInferenceOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) `operand`: tensor of floating-point type.
(I2) `scale`: 1-dimensional tensor of floating-point type.
(I3) `offset`: 1-dimensional tensor of floating-point type.
(I4) `mean`: 1-dimensional tensor of floating-point type.
(I5) `variance`: 1-dimensional tensor of floating-point type.
(I6) `epsilon`: constant of type `f32`.
(I7) `feature_index`: constant of type `si64`.
(C1) 0 <= `feature_index` < rank(`operand`).
(C2) `operand`, `scale`, `offset`, `mean`, `variance` and `result` have the
same element type.
(C3) size(`scale`) = `dim(operand, feature_index)`.
(C4) size(`offset`) = `dim(operand, feature_index)`.
(C5) size(`mean`) = `dim(operand, feature_index)`.
(C6) size(`variance`) = `dim(operand, feature_index)`.
(C7) `operand` and `result` have the same type.
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) `operand` is not a tensor of floating-point type. (Covered by ODS).
I2: a) `scale` is not a 1-dimensional tensor. (Covered by ODS).
    b) element_type(`scale`) != floating-point type. (Covered by ODS).
I3: a) `offset` is not a 1-dimensional tensor. (Covered by ODS).
    b) element_type(`offset`) != floating-point type. (Covered by ODS).
I4: a) `mean` is not a 1-dimensional tensor. (Covered by ODS).
    b) element_type(`mean`) != floating-point type. (Covered by ODS).
I5: a) `variance` is not a 1-dimensional tensor. (Covered by ODS).
    b) element_type(`variance`) != floating-point type. (Covered by ODS).
I6: a) `epsilon` is not a constant of type `f32`. (Covered by ODS).
I7: a) `feature_index` is not a constant of type `si64`. (Covered by ODS).
C1: a) `feature_index` < 0.
    b) `feature_index` >= rank(`operand`).
C2: a) element_type(`operand`) != element_type(`scale`). (Covered by ODS).
    b) element_type(`operand`) != element_type(`offset`). (Covered by ODS).
    c) element_type(`operand`) != element_type(`mean`). (Covered by ODS).
    d) element_type(`operand`) != element_type(`variance`). (Covered by ODS).
    e) element_type(`operand`) != element_type(`result`). (Covered by ODS).
C3: a) size(`scale`) != `dim(operand, feature_index)`.
C4: a) size(`offset`) != `dim(operand, feature_index)`.
C5: a) size(`mean`) != `dim(operand, feature_index)`.
C6: a) size(`variance`) != `dim(operand, feature_index)`.
C7: a) type(`operand`) != type(`result`).
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
C1a: `feature_index` < 0.
C1b: `feature_index` >= rank(`operand`).
C3a: size(`scale`) != `dim(operand, feature_index)`.
C4a: size(`offset`) != `dim(operand, feature_index)`.
C5a: size(`mean`) != `dim(operand, feature_index)`.
C6a: size(`variance`) != `dim(operand, feature_index)`.
C7a: type(`operand`) != type(`result`).
```

closes #963

### CODE REVIEW

File: stablehlo/dialect/TypeInference.cpp [hunk 1]
```diff
@@ -242,13 +244,15 @@ LogicalResult verifyBatchNorm(std::optional<Location> location,
   if (featureIndex < 0)
     return emitOptionalError(location, "expects featureIndex to be a ",
                              "non-negative number, got ", featureIndex, ".");
-  // Note: the above checks '0 <= feature-index < multiDimType.getRank()'
+  // Note: the above checks '0 <= feature_index < multiDimType.getRank()'
```
Comment (sdasgup3) [inline stablehlo/dialect/TypeInference.cpp:247]: Is this already existing comment relevant?

File: stablehlo/tests/ops_stablehlo.mlir [hunk 2]
```diff
@@ -4707,9 +4707,10 @@ func.func @error_batch_norm_train(%input: tensor<2x2x2x2xf32>, %scale: tensor<3x
 
 // CHECK-LABEL: @batch_norm_inference
```
Comment (sdasgup3) [inline stablehlo/tests/ops_stablehlo.mlir:4708]: add a test for c2

File: stablehlo/reference/Ops.cpp [hunk 3]
```diff
@@ -319,6 +335,35 @@ Tensor evalAndOp(const Tensor &lhs, const Tensor &rhs, TensorType resultType) {
...
+      evalBroadcastInDimOp(mean, {featureIndex}, operand.getType());
+  auto varianceBroadcast =
+      evalBroadcastInDimOp(variance, {featureIndex}, operand.getType());
+  auto epsilonBroadcast = evalBroadcastInDimOp(
+      makeTensor(DenseElementsAttr::get(
+          RankedTensorType::get({}, getF32Type(resultType.getContext())),
+          {epsilon})),
+      {}, operand.getType());
+
+  auto centeredOperand = evalSubtractOp(operand, meanBroadcast, resultType);
```
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:356]: for the eval functions here and below, I do not think we should use the `resultType` as the last argument,  but use the `getType()`  of any operand, say the first. 

I agree that result type of any of these evals must be equal to `resultType`, but that is an information we derived from the type inference of a series of associated operations flowing to the result.

File: stablehlo/dialect/StablehloOps.td [hunk 4]
```diff
@@ -1709,7 +1709,9 @@ def StableHLO_BatchNormGradOp : StableHLO_Op<"batch_norm_grad", [Pure,
 }
 
 def StableHLO_BatchNormInferenceOp : StableHLO_Op<"batch_norm_inference",
-    [Pure, AllElementTypesMatch<["operand", "result"]>, InferTensorType]> {
+    [Pure, AllElementTypesMatch<["operand", "scale", "offset", "mean",
+     "variance", "result"]> /*batch_norm_inference_c2*/,
+     InferTensorType /*batch_norm_inference_c7*/]> {
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/dialect/StablehloOps.td:1714]: This item was recently added to the checklist hence the discrepancy: https://github.com/openxla/stablehlo/blob/main/docs/reference_checklist.md?plain=1#L40

File: stablehlo/dialect/TypeInference.cpp (hunk 1 above)
Comment (ghpvnist) [reply stablehlo/dialect/TypeInference.cpp:247]: Not really, will remove.

File: stablehlo/reference/Ops.cpp (hunk 3 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:356]: Agreed.

**Review by sdasgup3 — APPROVED** [review]

File: stablehlo/reference/Ops.cpp [hunk 5]
```diff
@@ -323,6 +335,36 @@ Tensor evalAndOp(const Tensor &lhs, const Tensor &rhs, TensorType resultType) {
...
+      evalBroadcastInDimOp(scale, {featureIndex}, operand.getType());
+  auto offsetBroadcast =
+      evalBroadcastInDimOp(offset, {featureIndex}, operand.getType());
+  auto meanBroadcast =
+      evalBroadcastInDimOp(mean, {featureIndex}, operand.getType());
+  auto varianceBroadcast =
+      evalBroadcastInDimOp(variance, {featureIndex}, operand.getType());
+  auto epsilonBroadcast = evalBroadcastInDimOp(
+      makeTensor(DenseElementsAttr::get(
+          RankedTensorType::get({}, operand.getElementType()), {epsilon})),
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:352]: Nice catch! Let's change the spec to say `constant(epsilon, type(operand))` instead of just `constant(epsilon)`?

**Review by burmako — APPROVED** [review]
LGTM for the BatchNormInferenceOp part with one minor comment. Will review ConvertOp separately.

**Review by burmako — APPROVED** [review]

**Pull request merged**